update sequence diagrams based on feedback (#2)
redhat-et/ekho: sequence-diagrams/host-provisioning-xPU-VF.md

> 

Author: mention that the focus of this document will be the 2 cluster scenario

- Service metrics will be collectable through services that are fully integrated
as a Kubernetes deployment.

- The Kublet will be host networked.

> - The infrastructure is the `primary` entity driving/making networking decisions on behalf of the host.

Author: ```suggestion
- The infrastructure (AKA the xPU) is the `primary` entity driving/making networking decisions on behalf of the host.
```

## Opens

- If the infra cluster needs access to many Kubernetes resources in the tenant cluster, does it make sense to have a broker or is direct cluster access a better approach and does a broker architecture scale well across clusters when many of the resources that need to be shared are Kubernetes resources.
- Should we even consider using something like microshift on the DPU in the two cluster scenario.
- Need to think about how nodeport will work.
- How to use CNI in hetrogenous deployments.

### To Broker or not to Broker?

> A **Broker** is an entity that syncs OPI CRDs between the Tenant and Infra clusters. It's a singleton component that is deployed on a cluster whose Kubernetes API must be accessible by all of the participating clusters. The Broker cluster may be one of the participating clusters or a standalone cluster without the other components deployed. The Agent components deployed in each participating cluster are configured with the information to securely connect to the Broker cluster’s API.

Author: ```suggestion
A **Broker** is an entity that could be used to sync OPI CRDs between the Tenant and Infra clusters. It's a singleton component that is deployed on a cluster whose Kubernetes API must be accessible by all of the participating clusters. The Broker cluster may be one of the participating clusters or a standalone cluster without the other components deployed. The Agent components deployed in each participating cluster are configured with the information to securely connect to the Broker cluster’s API.
```

> The xPU Kubernetes architecture will avoid leveraging a Broker within the xPU as the infra cluster and the tenant cluster are tightly coupled. The infra cluster will need a access to several Kubernetes resources and it does not make sense to replicate these (resources) through CRDs via a Broker. There's also a question of latency introduced by the Broker and whether or not it can scale.

Author: Mark this as a proposal

An alternative option would be to eliminate the host agent, and the sequence diagram for that looks as follows:

![host-provisioning-initialization-no-host-agent](./images/host-provisioning-xPU-VF-init-no-host-agent.png)

#### Pod Creation

![host-provisioning-pod-creation](./images/host-provisioning-xPU-VF-pod-creation.png)

An alternative option would be to eliminate the host agent, and the sequence diagram for that looks as follows:

![host-provisioning-pod-creation-no-host-agent](./images/host-provisioning-xPU-VF-pod-creation-no-host-agent.png)

#### Pod Deletion

![host-provisioning-pod-deletion](./images/host-provisioning-xPU-VF-pod-deletion.png)

An alternative option would be to eliminate the host agent, and the sequence diagram for that looks as follows:

![host-provisioning-pod-deletion-no-host-agent](./images/host-provisioning-xPU-VF-pod-deletion-no-host-agent.png)

## CRDs

### Netdev

```golang
type NetdevStatus string

const (
    Connected NetdevStatus = "connected"
    Connecting NetdevStatus = "connecting"
    Disconnected NetdevStatus = "disconnected"
    ConnectionError NetdevStatus = "error"
)

type NetdevSpec struct {

>     NetdevID string `json:"netdev_id"` // SerialNum_PFID_SFID

Reviewer: Purely for reference, this is what is passed in annotations in OVN-Kubernetes (written as a map indexed by the namespace):
```
type DPUConnectionDetails struct {
	PfId         string `json:"pfId"`
	VfId         string `json:"vfId"`
	SandboxId    string `json:"sandboxId"`
	VfNetdevName string `json:"vfNetdevName,omitempty"`
}
```

The IP info was already in the Pod Annotation, so it is not part of the DPU specific content.
```
// Internal struct used to marshal PodAnnotation to the pod annotation
type podAnnotation struct {
	IPs      []string   `json:"ip_addresses"`
	MAC      string     `json:"mac_address"`
	Gateways []string   `json:"gateway_ips,omitempty"`
	Routes   []podRoute `json:"routes,omitempty"`

	IP      string `json:"ip_address,omitempty"`
	Gateway string `json:"gateway_ip,omitempty"`
}

// Internal struct used to marshal PodRoute to the pod annotation
type podRoute struct {
	Dest    string `json:"dest"`
	NextHop string `json:"nextHop"`
}
```

Reviewer: I forgot to mention, for IP, if I recall, `ip_address` and `gateway_ip` were there originally, and then when dual stack was added, it became an array of address, `ip_addresses` and `gateway_ips`. The original was left around for backwards compatibility.

    Hostname string `json:"hostname"`
    Subnets []string `json:"subnets"` // or maybe other routing info?
    IP string `json:"ip"`
    Backend string `json:"backend,omitempty"` //subfunction or virtual function
    Status NetdevStatus `json:"status"`
}

type Netdev struct {
    metav1.TypeMeta `json:",inline"`
    metav1.ObjectMeta `json:"metadata,omitempty"`

>     Spec NetdevSpec `json:"spec"`

Reviewer: Probably want status as a separate structure, that way Host owns Spec and DPU owns Status. Question is whether IP belongs in Spec or Status (depending on whether Host or DPU writes it).
```suggestion
    Spec NetdevSpec `json:"spec"`
    // +optional
    Status NetdevStatus `json:"status,omitempty"`
```

Author: makes sense

  - Configures the interface with the allocated IP address
  - Moves the VF/SF from the Host network namespace to the Pod Network namespace
   (vice versa).

### Required Resources

> TODO (WIP)

Author: This is a non exhaustive list of resources that need to be shared between a tenant and infra cluster

- Netdev CRD (New)
- Service: reflecting the services in the tenant cluster.
- EndpointSlice: reflecting the endpoint slices in the tenant cluster.
- Network policy: reflecting the network policies in the tenant cluster.

### Single Cluster sequence diagrams

TODO

### Multi Cluster sequence diagrams

#### Initialization

The following high level diagram provides an overview of the initialization of the entities involved in provisioning an xPU VF to a host.

![host-provisioning-initialization](./images/host-provisioning-xPU-VF-init.png)

The following high level diagram provides an overview of the entities working
together to provision a secondary network interface for a pod.

> TODO

Author: I'm not sure this looks different to the previous diagram
